### ISSUE TITLE
Up axis extension

### ISSUE THREAD

**Issue by lilleyse** [KhronosGroup/glTF#1324]
Would there be any interest in an extension that allows redefining which axis is the up axis?

I come to this from a 3D geospatial angle where globe engines and 3D models often use the WGS84 coordinate system, whose origin is the Earth's center and up-axis is the north pole - essentially a z-up coordinate system.

In order to bring WGS84 (inherently z-up) models into Cesium we have to go through a few hoops including attaching a z-up to y-up transform to the root node of the glTF and then applying a reverse y-up to z-up transform at runtime. Related to this, we are building [3D Tiles](https://github.com/AnalyticalGraphicsInc/3d-tiles) with a z-up coordinate system and it is confusing that the glTF payload is y-up.

To summarize, it would be convenient if we could specify a different up axis than y.

The obvious downside to an extension such as this is now different glTF models are not following the same convention. Nonetheless would this extension be of interest to others?

**Comment by ggetz** [comment]
Additionally, geospatial is not the only use case for z-up. Other 3D applications such as 3DS Max and Unreal engine are z-up.

**Comment by lilleyse** [comment]
I imagined it would be implemented by applying an additional transform to the node hierarchy, similar to how a z-up engine would load a y-up model today, and not make changes to binary contents.

**Comment by significant-bit** [comment]
Please no. One of the strengths of glTF is that it has one particular coordinate system & units.

I make the importer for Unreal (z up) and am happy that every file has the same up direction, so I only have to get that part of the code right **once**! 😄 

Making this more flexible only invites trouble.

**Comment by significant-bit** [comment]
* that said, Unreal is not importing directly from BIN to VRAM like some runtime engines might. We transform the data quite a bit before using it, so adding axis conversion is not a big deal.

**Comment by recp** [comment]
> I imagined it would be implemented by applying an additional transform to the node hierarchy, similar to how a z-up engine would load a y-up model today, and not make changes to binary contents.

Actually I implemented converting UP axis in [AssetKit](https://github.com/recp/assetkit) for COLLADA and it can convert any coord-sys to another, even if exernal COLLADA or some elements in COLLADA file may have different UP axis, all will be fixed (optional). 

Converting one glTF file from Y_UP to Z_UP (or to other) wouldn't be a problem for single glTF file, applying transforms to nodes would fix UP axis. 

But as I mentioned before, if two or more glTF files share same binary or other external files, there will be up axis conflicts between glTF files for BINs. We cannot apply transforms to nodes here because we don't know BIN's up axis and we don't know which glTF file has ownership of shared BIN files 
(we could use this info to fix up axis).

If glTF will support multiple UP axis with extension or in spec, every case must be well defined. Maybe there will be other issues than above cases/examples

> One of the strengths of glTF is that it has one particular coordinate system & units.



> Making this more flexible only invites trouble.

+1

**Comment by donmccurdy** [comment]
I would prefer to avoid implementing support for other coordinate systems in three.js — consistency of the coordinate system and units is valuable to client implementations, and lack of that consistency has been a pain point with other formats.

For the use case described above, something like a vendor-prefixed `CESIUM_coordinates` extension seems preferable to a more general extension. Another option would be some semantic way of indicating that a node has been introduced by tooling rather than a user decision — e.g. marking it in extras or an extension as disposable, and allowing further tools to flatten the scene graph and dispose of that node when converting to another coordinate system without fear of losing information the user wanted to keep.

**Comment by lilleyse** [comment]
Ok, I think there is pretty good consensus here that a general extension will add too much complexity. A `CESIUM` prefixed extension may be the right way to go.

**Comment by lilleyse** [comment]
Thanks for the feedback everyone.

For the geospatial/3D Tiles use case we've decided to stick with the default glTF y-up. If anyone is interested we laid out some notes for this use case here: https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/1.0#gltf.
